### DISCUSSION TITLE
0.99.2.0 for MC 1.16.* & 1.17.* & 1.18.* & 1.19.* & 1.20.*

### DISCUSSION THREAD

**Discussion by LlmDl** [TownyAdvanced/Towny#6763] [Release Discussions]
# Towny Advanced 0.99.2.0

<p><img align=left style="padding: 40px;" src="https://user-images.githubusercontent.com/879756/65964696-19d6b300-e423-11e9-9cb0-d193225ee40f.png">

### Release contains:
Towny 0.99.2.0
TownyChat 0.107
<br>

|Welcome to Towny 0.99.2.0.|
| ---- | 
| Towny 0.99.2.0 is the third release in the 0.99.\*.\* series! |
| 0.99.2.0 supports all versions of Minecraft 1.16 and newer. | 
| This version adds support for MC 1.20.*, a Nation conquered tax, Town Leaderboard PlaceholderAPI placeholders and more. |
| This version contains <strong>critical fixes</strong> that you want on your server! |
| <strong>This is a drag-and-drop update:</strong> just replace your old .jars with the new ones and Towny will take care of the rest of the update by itself. |

<details open>
<summary> ➕ Added (14)</summary><blockquote>

## ➕ Added:

  - Added Nation Conquered Tax.
    - A feature that lets nations set a conquered tax which is paid by the conquered towns, in addition to the normal nation tax.
    - Conquered Tax is set using /n set conqueredtax {amount}
    - ConqueredTax appears on the nation status screen.
    - Closes #6397.
  - Add Town leaderboard PAPI Placeholders.
    - Closes #5073.
  - Add option to allow enemies to spawn to neutral/peaceful towns.
    - Closes #6401.
  - Add towny.command.resident.friend to the nomad group for newly made Townyperms.yml files.
  - Refactor upkeep & taxation code, improving the messages shown to the players about how much tax is collected as well as what appears in the bank history books.

</blockquote>
<blockquote>

  - 1.20.* Support, courtesy of Warrior with PR #6739.
    - Protect Suspicious Sand and Gravel being brushed, using destroy tests.
    - Update readme.md with version info.
  - Add persistent cooldowns, courtesy of Warrior with PR #6723.
    - Closes #2949.
  - Minor improvements to player cache, courtesy of Warrior with PR #6717.
  - Add default Geyser username compatability, courtesy of Warrior with PR #6724.
  - Port ItemList changes from enum branch, courtesy of Warrior with PR #6737.
  - Allow villagers to open doors, courtesy of Warrior with PR #6738.
  - Replace all printStackTraces with the logger, courtesy of Warrior with PR #6718.
  - Add new excludeTag method, courtesy of Warrior with PR #6756.
  - Use paper's isWaxed API for signs, courtesy of Warrior with PR #6744.

</blockquote>

----
</blockquote></details>

<details>
<summary>➕ Recommended Config Edits (1)</summary><blockquote>

## ➕ Towny's default config has had one change that you might want to make.
  - In your config.yml's switch_ids, change PRESSURE_PLATES to NON_WOODEN_PRESSURE_PLATES.

  - Default Config Change:
    - The switch_ids list will now come with the new NON_WOODEN_PRESSURE_PLATES as default, replacing PRESSURE_PLATES.
    - Wooden pressure plates are inherently un-protectable due to being set off by dropped items.
    - Wooden pressure plates are good for contraptions you would want mobs to use, or for doors which you want players to be able to use.
    - Closes #6752.

----
</blockquote></details>

<details>
<summary>✨ Command Changes (1)</summary><blockquote>

<details>
<summary>✨ New Commands (1)</summary>

## ✨ New Commands:

  - /nation set conqueredtax {amount}
    - Sets a nation's conquered tax, which is paid by the conquered towns.

  
----
</details>

<details>
<summary>✨ Changed Commands (0)</summary>

## ✨ Changed Commands:

  - None.
  
----
</details>

<details>
<summary>❌ Removed Commands (0)</summary>

## ❌ Removed Commands:
  - None.

----
</details>

----
</blockquote></details>

<details>
<summary>📕 Config Changes (4)</summary><blockquote>

<details>
<summary> 📕 New Config Options (4)</summary>

#### These new settings will be added to your config with their default settings.
  - nation.default_taxes.default_nation_conquered_tax
    - Default: 0
    - The default amount of money that nations will charge their conquered towns.

  - nation.default_taxes.max_nation_conquered_tax
    - Default: 100
    - The maximum amount of money that can be charge by a nation on their conquered towns.

  - filters_colour_chat.papi_leaderboard_formatting.format"
    - Default: "%s - %s",
    - How the %townyadvanced_top_....% placeholders will appear, first %s being the town name, with the second being the balance, resident count or town size.

  - global_town_settings.allow_enemies_spawn_to_peaceful_towns
    - Default: false
    - When true, players will be allowed to spawn to peaceful/neutral towns in which they are considered enemies.
    - Setting this to true will make town spawn points unsafe for private towns which are part of nations with enemies.

----
</details>

<details>
<summary>❌ Removed Config Options (1)</summary>

## ❌ Removed Config Options:
  - notification.owner_shows_nation_title
  
----
</details>

----
</blockquote></details>

<details>
<summary>🔑 Permission Node Changes (1)</summary><blockquote>

<details>
<summary>🔑 New Permission Nodes (1)</summary>

## 🔑 New Permission Nodes:
  **No change required in townyperms.yml.** 
  - towny.command.nation.set.conqueredtax
    - Child node of towny.command.nation.set.*
    - No change required to the townyperms.yml

----
</details>

<details>
<summary>❌ Removed Permission Nodes (0)</summary>

## ❌ Removed Permission Nodes:
  - None.

----
</details>
----
</blockquote></details>

<details>
<summary>💻 API Changes (1)</summary><blockquote>

## 💻 API: 
  - Added TownPaysNationConqueredTaxEvent
    - Thrown before Towny charges a conquered town the nation conquered tax.
    - A cancellable event that lets a plugin alter the amount if they choose.

</blockquote>

----
</blockquote></details>

<details>
<summary>🔧 Fixed (23)</summary><blockquote>
  
## 🔧 Fixed: 

  - Fix regression dating to 0.99.1.1: some of the town and nation lists being in reverse order.
    - Closes #6746.
  - Fix nation ranking issue.
    - Closes #6762.
  - Fix merge costs related to purchaseable bonus blocks.
  - Fix ancient typo in the plugin.yml's towny.command.town.* childpermissions nodes.

</blockquote>
<blockquote>

  - Fix changelog warning message when downgrading, courtesy of Warrior with PR #6741.
    - Changelog will only be printed out when the version is newer.
  - Fix the seventh null mayor npe, courtesy of Warrior with PR #6710.
  - Fix exception when using negative radius for a circle claim, courtesy of Warrior with PR #6712.
  - Fix legacy debt converter not respecting isEconomyAsync, courtesy of Warrior with PR #6716.
  - Use a concurrent hashmap for invite handler invites, courtesy of Warrior with PR #6719.
  - Remove 'install vault' recommendation if it's present, courtesy of Warrior with PR #6715.
  - Replace last use of non-get player exact, courtesy of Warrior with PR #6725.
  - Skip is warmup task running for movement, courtesy of Warrior with PR #6729.
  - Add reload listeners to some timer tasks, courtesy of Warrior with PR #6730.
  - Fix cooldown loading exception message on first startup, courtesy of Warrior with PR #6736.
  - Cleanup the mob removal task, courtesy of Warrior with PR #6740.
  - Fix possible reclaimTown NPE, courtesy of Warrior with PR #6753.

</blockquote>

----    
</blockquote></details>

<details>
<summary>🌏 Language Updates (0)</summary><blockquote>

## 🌏 Language Updates:
  - None.

----
</blockquote></details>

<details>
<summary>🌏 Dependency Updates (3)</summary><blockquote>

## 🌏 Dependency Updates:
  - Bump maven-surefire-plugin from 3.1.0 to 3.1.2.
  - Bump adventure-text-serializer-plain from 4.13.1 to 4.14.0.
  - Bump adventure-text-minimessage from 4.13.1 to 4.14.0.
  
----
</blockquote></details>

<details>
<summary>👽 Placeholder Updates (5)</summary><blockquote>

## 👽 Placeholder Updates:
  - `%townyadvanced_resident_primary_rank%` - Displays a prefix associated with a player's highest-priority nation or town rank.
  - `%townyadvanced_resident_primary_rank_spaced%` - Displays a resident's primary rank with a space afterwards.
  - `%townyadvanced_town_map_color_minimessage_hex%` - Shows the town colour formatted for MiniMessage, (TownyChat-friendly.)
  - `%townyadvanced_nation_map_color_minimessage_hex%` - Shows the nation colour formatted for MiniMessage, (TownyChat-friendly.)
  - `%townyadvanced_top_town_balance_n%` - (Replace n with a number) Displays the nth Town by money in the town bank.
  - `%townyadvanced_top_town_residents_n%` - (Replace n with a number) Displays the nth Town by number of residents.
  - `%townyadvanced_top_town_land_n%` - (Replace n with a number) Displays the nth Town by number of townblocks claimed.

----
</blockquote></details>

<details>
<summary>💬 TownyChat Changes (2)</summary><blockquote>
  
## 💬 TownyChat Changes: 

  - Remove usage of Version, using new Towny version checker instead.
  - Handle knock-off Essentials plugins fooling TownyChat into hooking the real EssentialsX.

----
</blockquote></details>

<table align=center>
<tr><th colspan=6><h3>Towny Sponsors<br><h4>I want to give a big thank you to all of my sponsors:<br>These are the people that help to make Towny's support and development as active as it is,<br>and who supported me during this pre-release of Towny. </th></tr>
<tr><td>MrLogano</td><td>Momshroom</td><td>Cobrex1</td><td>TheFlagCourier</td><td>DrprofLuigi</td><td>SmallSansSerif</td></tr>
<tr><td>TheNylox</td><td>pebblehost</td><td>NoikzGaming</td><td>ThePontificate</td><td>AllieDragon</td><td>TegulaThePig</td></tr>
<tr><td>darthpeti</td><td>Eblyss</td><td>akup001</td><td>Valerie2308</td><td>Kelbbie</td><td>tjdings</td></tr>
<tr><td>AldinaMapping</td><td>CrypticBlades</td><td>EcoSMP</td><td>Semisol</td><td>Theeno74</td><td>ItzXPurpleX</td></tr>
<tr><td>Rekar7</td><td>Noxillio</td><td>notzune</td><td>Lambsy</td><td>guillermodiazkiernans</td><td>ItzPear</td></tr>
<tr><td>DeltaOrion </td><td>Tobza</td><td>vyvin</td><td>keijodputt</td><td>ArmonicPvP</td><td>NeumimTo</td></tr>
<tr><td>o-zzer</td><td>Dracathio</td><td>Peskdt</td><td>Merlin6019 </td><td>vette-daniel</td><td>AncientSM</td></tr>
<tr><td>JunkieCraft</td><td>CalebGitBritannia</td><td>Hikitsune-Red</td><td>DoctorChosen</td><td>Dynamite99</td><td>ItsSynchro</td></tr>
<tr><td>EpicSpeed</td><td>SuperRedstoneYT</td><td>Jayoakson</td><td>tyhppp233</td><td>MrLogano</td><td>Tuaniwan</td></tr>
<tr><td>CorruptedGreed</td><td>ProjectGuinness</td><td>Deathrayne225</td><td>EBatuzay</td><td>Khargot</td><td>Mineprank</td></tr>
<tr><td>ReEvadere</td><td>trevor1097</td><td>MetaverseRP</td><td>CorruptVD</td><td>GabrielTK</td><td>bretomart</td></tr>
<tr><td>Polo-lulu</td><td>Folas1337</td><td>GundalfWasTaken</td><td>qwlke</td><td>MayorMike</td><td>Draedixe</td></tr>
<tr><td>gageeet</td><td>wizard1233</td><td>Th35p1k3</td><td>j0g32</td><td>Wextra</td><td>Microweb4</td></tr>
<tr><td>tskopen</td><td>PainOchoco </td><td>Bloc-k</td><td>Radiation-pi</td><td>Aviside</td><td>bradyyykennyy</td></tr>
<tr><td>xSTORMYxSHM</td><td>StevenSharker</td><td>Manul-Newton</td><td>Polo-lulu</td><td>Ashparty </td><td>jsuarezl </td></tr>
<tr><td>royblue12</td><td>mikslendov</td><td>BeepSterr</td><td>nhannon9</td><td>ItzMeKarlix</td><td>tapok233</td></tr>
<tr><td>tomhmagic</td><td>WyattPereira</td><td>o7-hmoe </td><td>Rialkingg</td><td>Roman90879</td><td>editon96 </td></tr>
<tr><td>Omic104</td><td>Intrabit</td><td>Leevos</td><td>welenwat</td><td>gerardvanschip </td><td>TheWordSmith123</td></tr>
<tr><td>GowtherSukehiro</td><td>JackTheYack</td><td colspan=4>and 36 private sponsors.</td></tr>
<tr><td colspan=6 align=center>
<h3><a href=https://github.com/sponsors/LlmDl>If you want to support the developer, become a sponsor.</a></h3>
(It's just like Patreon but instead 100% of your support goes to the developer.)
<h3>Plus there are <a href=https://github.com/LlmDl/SponsorPlugins/blob/main/README.md>Sponsor Plugins!</a></h3></td></tr>
</table>

<table align=center>
<th colspan=4> Important Links </th>
<tr align=center>
<td><a title="Frequently Asked Questions" href=https://github.com/TownyAdvanced/Towny/wiki/Frequently-Asked-Questions>Frequently Asked<br>Questions</a></td>
<td><a title="How Towny Works" href=https://github.com/TownyAdvanced/Towny/wiki/How-Towny-Works>How Towny Works</a></td>
<td><a title="Install Guide" href=https://github.com/TownyAdvanced/Towny/wiki/Installation>Towny Install Guide</a></td>
<td><a title="Updating Towny" href=https://github.com/TownyAdvanced/Towny/wiki/Updating-Towny>Towny Update Guide</a></td></tr>
<tr align=center>
<td><a title="Other Towny Plugins" href="https://townyadvanced.github.io" target="_blank" rel="noopener">Other Towny Plugins</a> </td>
<td><a title="Towny Changelog" href="https://github.com/TownyAdvanced/Towny/blob/master/Towny/src/main/resources/ChangeLog.txt" target="_blank" rel="noopener">Complete Changelog</a> </td>
<td><a title="Default Config Files" href="https://github.com/TownyAdvanced/Towny/wiki/Config-Files">Default Config Files</a></td>
<td><a title="Commands/Permissions/Placeholders" href="https://github.com/TownyAdvanced/Towny/wiki/Reference">Commands/Permissions<br>Placeholders</a></td></tr>
</table>

### ⏬ Download available as a .zip file in the Assets section below:

<hr /><em>This discussion was created from the release <a href='https://github.com/TownyAdvanced/Towny/releases/tag/0.99.2.0'>0.99.2.0 for MC 1.16.* & 1.17.* & 1.18.* & 1.19.* & 1.20.*</a>.</em>

**Comment by OltsuX** [comment]
how can i get this a jar file?

**Reply by LlmDl** [reply]
The release page: https://github.com/TownyAdvanced/Towny/releases/tag/0.99.2.0

At the bottom.
